lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -283,58 +283,75 @@ static bool isBound(hw::HWModuleLike op, hw::InstanceGraph &instanceGraph) {
-  }
-
-  // Move the module to the test code path.
-  oldMod->setAttr("output_file", testBenchDir);
-
-  // Replace its instances with binds.
+  // Iterate through each of the module's inputs.
   hw::InstanceGraphNode *node = instanceGraph.lookup(oldMod);
-  OpBuilder b = OpBuilder(oldMod.getContext());
+  OpBuilder b(oldMod);
+  bool allInlined = true;
   for (hw::InstanceRecord *use : node->uses()) {
-    OpBuilder::InsertionGuard g(b);
-
+    // If there is no instance, move on.
     hw::HWInstanceLike instLike = use->getInstance();
-    if (!instLike)
+    if (!instLike) {
+      allInlined = false;
       continue;
+    }
 
-    // Mark the instance to not be printed.
+    // If the instance had a symbol, we can't inline it.
     hw::InstanceOp inst = cast<hw::InstanceOp>(instLike.getOperation());
-    hw::HWModuleLike parent = use->getParent()->getModule();
-    inst->setAttr("doNotPrint", b.getBoolAttr(true));
-
-    // Give the instance an inner sym if it needs one.
-    if (!inst.getInnerSym().has_value())
-      inst.setInnerSymAttr(
-          b.getStringAttr("__" + inst.getInstanceName() + "__"));
-
-    // Create the bind.
-    b.setInsertionPointToEnd(
-        &oldMod->getParentOfType<mlir::ModuleOp>()->getRegion(0).front());
-    auto bindOp = b.create<sv::BindOp>(inst.getLoc(), parent.moduleNameAttr(),
-                                       inst.getInnerSymAttr());
-    if (bindFile)
-      bindOp->setAttr("output_file", bindFile);
+    if (inst.getInnerSym().has_value()) {
+      allInlined = false;
+      continue;
+    }
+
+    // Build a mapping from module block arguments to instance inputs.
+    BlockAndValueMapping mapping;
+    assert(inst.getInputs().size() == oldMod.getNumInputs());
+    auto inputPorts = oldMod.getBodyBlock()->getArguments();
+    for (size_t i = 0, e = inputPorts.size(); i < e; ++i)
+      mapping.map(inputPorts[i], inst.getOperand(i));
+
+    // Inline the body at the instantiation site.
+    hw::HWModuleOp instParent =
+        cast<hw::HWModuleOp>(use->getParent()->getModule());
+    OpBuilder::InsertionGuard g(b);
+    b.setInsertionPoint(inst);
+    for (auto &op : llvm::make_early_inc_range(*oldMod.getBodyBlock())) {
+      // For instances in the bind table, update the bind with the new parent.
+      if (auto innerInst = dyn_cast<hw::InstanceOp>(op)) {
+        if (auto innerInstSym = innerInst.getInnerSymAttr()) {
+          auto it = bindTable[oldMod.getNameAttr()].find(innerInstSym);
+          if (it != bindTable[oldMod.getNameAttr()].end()) {
+            sv::BindOp bind = it->second;
+            auto oldInnerRef = bind.getInstanceAttr();
+            auto newInnerRef = hw::InnerRefAttr::get(
+                instParent.moduleNameAttr(), oldInnerRef.getName());
+            bind.setInstanceAttr(newInnerRef);
+          }
+        }
+      }
+
+      // For all ops besides the output, clone into the parent body.
+      if (!isa<hw::OutputOp>(op))
+        b.clone(op, mapping);
```

Comment:
Do we have any ops with regions that we have to worry about at the HW level? I guess the worry is that an instance op could live in some random operation.

HW is a graph region, does clone work correctly for graph regions?
```
%0 = comb.add %1, %1
%1 = comb..add %0, %0
```

Reply:
This will clone regions over and take care of remapping things with the `mapping`. The non-region version is cloneWithoutRegions, but this should work for sv.if and friends. I don't know if it's possible to have an instance inside a region, but if that happened it should be moved over by this.

Regarding cycles, I guess that could be an issue for `clone`, but I haven't seen it show up in practice. I'm not sure a full comb cycle detection is a good idea here, so I don't really know what to do. In practice, the kinds of modules this inlines feedforward any combinational or sequential logic directly into the extracted test code instance. Is there a cheap way to check that?